fixes #13309

planvin opened the pull request:
## Summary & Motivation
An attempt to fix #13309
## How I Tested These Changes

patched impl.py locally

Launched a grpc instance as described in issue #13309
Launched a dagit instance as described in issue #13309

Materialization launched normally

gibsondan commented:
Thanks so much for sending this out! This looks right to me, we’ll take a closer look on Monday.

smackesey commented on python_modules/dagster/dagster/_grpc/impl.py:
```diff
@@ -156,6 +156,18 @@ def core_execute_run(
 
 
+def _instance_from_ref(instance_ref: InstanceRef, partitions_def: PartitionsDefinition):
+    # Certain gRPC servers do not have access to the instance, so we only attempt to instantiate
```
- A helper method here is reasonable, but let's call it `_instance_from_ref_for_dynamic_partitions` to communicate that it's not always yielding an instance.
- This needs a `@contextmanager` decorator and the result yielded instead of returned to work as intended
- Annotate return vale as `Iterator[Optional[DagsterInstance]]`
- This should never be called with a `partitions_def=None`, so no need to check

planvin replied:
thanks, applied these changes in last commit

smackesey commented on python_modules/dagster/dagster/_grpc/impl.py:
```diff
@@ -525,14 +531,7 @@ def get_partition_set_execution_param_data(
-            else None
-        )
-        with user_code_error_boundary(
+        with _instance_from_ref(instance_ref, partitions_def) as instance, user_code_error_boundary(
```
Let's put the `user_code_error_boundary` nested on a separate line

clairelin135 commented on python_modules/dagster/dagster/_grpc/impl.py:
```diff
@@ -401,24 +413,19 @@ def get_partition_config(
-        else:
-            partition = partitions_def.get_partition(partition_key, dynamic_partitions_store=None)
+        with _instance_from_ref(instance_ref, partitions_def) as instance:
+            partition = partitions_def.get_partition(
```
I think the `get_partition` call is unnecessary, cc @smackesey to confirm

I think we fetch the partition object within `get_run_config_for_partition_key` to validate that it is a valid partition

smackesey replied:
I think you're right, but it's not returned, while we need the partition name out here for return in the `ExternalPartitionConfigData`.

clairelin135 replied:
I think the partition key is always partition.name

smackesey replied:
You're probably right but I think you should just remove it in your followup PR that already exists to avoid adding complexity and potential back-and-forth with @planvin, since these `get_partition` calls pre-existed his PR anyway.

clairelin135 replied:
👍

clairelin135 commented on python_modules/dagster/dagster/_grpc/impl.py:
```diff
@@ -462,20 +469,19 @@ def get_partition_tags(
-                partition.name, job_name=job_def.name
+
+        with _instance_from_ref(instance_ref, partitions_def) as instance:
+            partition = partitions_def.get_partition(
```
same here

smackesey commented:
@planvin thx, running our CI against this now

smackesey commented on python_modules/dagster/dagster/_grpc/impl.py:
```diff
@@ -401,24 +415,19 @@ def get_partition_config(
+                lambda: f"Error occurred during the evaluation of the `run_config_for_partition` function for partition set {partition_set_name}",
+            ):
+                run_config = partitioned_config.get_run_config_for_partition_key(
+                    partition_key, instance=instance
```
`instance=instance` -> `dynamic_partitions_store=instance`

smackesey commented:
@planvin CI shows some failures:

- instance=instance -> dynamic_partitions_store=instance (see my inline comment)
- You need to change the annotation of `instance_ref` in your helper to `Optional[InstanceRef]`, and yield null context if `instance_ref` is None
- Run `make ruff` and make sure it passes
- Rebase on master before repush

Hopefully those will get this to green

planvin commented:
> @planvin CI shows some failures:
> 
>     * instance=instance -> dynamic_partitions_store=instance (see my inline comment)
fixed
> 
>     * You need to change the annotation of `instance_ref` in your helper to `Optional[InstanceRef]`, and yield null context if `instance_ref` is None
fixed the typing. not sure about the yielding of nullcontext, if instance_ref is none we will yield None so I think we are ok here right ?
> 
>     * Run `make ruff` and make sure it passes
> 
ok
>     * Rebase on master before repush
> 
ok
> 
> Hopefully those will get this to green

smackesey commented:
Sweet, thx @planvin, merging this now!
